# HG changeset patch
# User Rik <rik@octave.org>
# Date 1541437428 28800
#      Mon Nov 05 09:03:48 2018 -0800
# Node ID f70729091570970e05d6fb8f0fae09dc9f6ac04f
# Parent  18449ef91387ff0d935b69118192966a06cf0a32
msgbox.m: Install "keypress" callback that closes box when <RETURN> pressed (bug #54942).

* msgbox.m (cb_callback): New callback function that closes figure when OK
button clicked.
* msgbox.m (cb_keypress): New callback function that closes figure when OK
button has focus and <RETURN> typed.
* msgbox.m: Connect uicontrol OK pushbutton callbacks to new callback
functions.

diff --git a/scripts/gui/msgbox.m b/scripts/gui/msgbox.m
--- a/scripts/gui/msgbox.m
+++ b/scripts/gui/msgbox.m
@@ -259,23 +259,37 @@ function hf = __msgbox__ (msg, tit, icon
   set (hf, "position", [center(1)-wd/2 center(2)-hg/2 wd hg]);
 
   set (hax, "position", [0 button_margin ax_sz], ...
        "xlim", [1 ax_sz(1)]-ax_margin, "ylim", [-ax_sz(2)/2 ax_sz(2)/2], ...
        "units", "normalized");
 
   set (ht, "units", "data", "position", [text_offset 0 0]);
 
-  hui = uicontrol ("string", "OK", "callback", @() close (gcbf ()), ...
+  hui = uicontrol ("string", "OK",
+                   "callback", @cb_callback,
+                   "keypressfcn", @cb_keypress,
                    "position", [ax_sz(1)/2-40 ax_margin 80 28], "parent", hp);
 
   set (hf, "windowstyle", windowstyle, "visible", "on");
 
 endfunction
 
+## Callback when button clicked (close window)
+function cb_callback (~, ~)
+  close (gcbf ());
+endfunction
+
+## Callback when key typed (close window on <RETURN>)
+function cb_keypress (~, ev)
+  if (strcmp (ev.Key, "return"))
+    close (gcbf ());
+  endif
+endfunction
+
 
 %!demo
 %! msgbox ("A bare dialog");
 
 %!demo
 %! msgbox ("An informative string", "Documentation", "help");
 
 %!demo
